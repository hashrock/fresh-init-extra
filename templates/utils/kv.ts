export interface Field {
  name: string;
  type: string;
  nullable: boolean;
}

export function typeCode(typeName: string, fields: Field[]) {
  const fieldsCode = fields
    .map((i) => `  ${i.name}${i.nullable ? "?" : ""}: ${i.type}`)
    .join(";\n");

  return `
export interface ${typeName} {
  id: string;
${fieldsCode}
  createdAt: Date;
  updatedAt: Date;
}`;
}

export function listCode(typeName: string) {
  const lower = typeName.toLowerCase();

  return `
export async function list${typeName}() {
  const iter = await kv.list<${typeName}>({ prefix: ["${lower}"] });
  const ${lower}: ${typeName}[] = [];
  for await (const item of iter) {
    ${lower}.push(item.value);
  }
  return ${lower};
}`;
}

export function updateCode(typeName: string, fields: Field[]) {
  const lower = typeName.toLowerCase();
  const argsCode = fields
    .map((i) => `${i.name}${i.nullable ? "?" : ""}: ${i.type}`)
    .join(", ");
  const assignCode = fields
    .map((i) => `  ${lower}.${i.name} = ${i.name};`)
    .join("\n");
  return `
export async function update${typeName}(id: string, ${argsCode}) {
  const ${lower} = await get${typeName}(id);
  if (!${lower}) throw new Error("${lower} not found");
${assignCode}
  ${lower}.updatedAt = new Date();
  await kv.set(["${lower}", id], ${lower});
}`;
}

export function deleteCode(typeName: string) {
  const lower = typeName.toLowerCase();

  return `
export async function delete${typeName}(id: string) {
  await kv.delete(["${lower}", id]);
}`;
}

export function getCode(typeName: string) {
  const lower = typeName.toLowerCase();
  return `
export async function get${typeName}(id: string) {
  const res = await kv.get<${typeName}>(["${lower}", id]);
  return res.value;
}`;
}

export function addCode(typeName: string, fields: Field[]) {
  const lower = typeName.toLowerCase();
  const argsCode = fields
    .map((i) => `${i.name}${i.nullable ? "?" : ""}: ${i.type}`)
    .join(", ");
  const fieldCode = fields.map((i) => `    ${i.name}`).join(",\n");

  return `
export async function add${typeName}(${argsCode}) {
  const id = crypto.randomUUID();
  const ${lower}: ${typeName} = {
    id,
${fieldCode},
    createdAt: new Date(),
    updatedAt: new Date(),
  };
  await kv.set(["${lower}", id], ${lower});
  return id;
}`;
}
