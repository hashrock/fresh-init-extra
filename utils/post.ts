const kv = await Deno.openKv();

export interface Post {
  id: string;
  title: string;
  body: string
  createdAt: Date;
  updatedAt: Date;
}

export async function listPost() {
  const iter = await kv.list<Post>({ prefix: ["post"] });
  const post: Post[] = [];
  for await (const item of iter) {
    post.push(item.value);
  }
  return post;
}

export async function addPost(title: string, body: string) {
  const id = crypto.randomUUID();
  const post: Post = {
    id,
    title,
    body,
    createdAt: new Date(),
    updatedAt: new Date(),
  };
  await kv.set(["post", id], post);
  return id;
}

export async function getPost(id: string) {
  const res = await kv.get<Post>(["post", id]);
  return res.value;
}

export async function updatePost(id: string, title: string, body: string) {
  const post = await getPost(id);
  if (!post) throw new Error("post not found");
  post.title = title;
  post.body = body;
  post.updatedAt = new Date();
  await kv.set(["post", id], post);
}

export async function deletePost(id: string) {
  await kv.delete(["post", id]);
}