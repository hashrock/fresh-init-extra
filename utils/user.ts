const kv = await Deno.openKv();

export interface User {
  id: string;
  title: string;
  body: string
  createdAt: Date;
  updatedAt: Date;
}

export async function listUser() {
  const iter = await kv.list<User>({ prefix: ["user"] });
  const user: User[] = [];
  for await (const item of iter) {
    user.push(item.value);
  }
  return user;
}

export async function addUser(title: string, body: string) {
  const id = crypto.randomUUID();
  const user: User = {
    id,
    title,
    body,
    createdAt: new Date(),
    updatedAt: new Date(),
  };
  await kv.set(["user", id], user);
  return id;
}

export async function getUser(id: string) {
  const res = await kv.get<User>(["user", id]);
  return res.value;
}

export async function updateUser(id: string, title: string, body: string) {
  const user = await getUser(id);
  if (!user) throw new Error("user not found");
  user.title = title;
  user.body = body;
  user.updatedAt = new Date();
  await kv.set(["user", id], user);
}

export async function deleteUser(id: string) {
  await kv.delete(["user", id]);
}