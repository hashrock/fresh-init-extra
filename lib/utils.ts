import * as kv from "../templates/utils/kv.ts";

export function generateExampleFromFields(fields: kv.Field[]) {
  const example: Record<string, unknown> = {};
  for (const field of fields) {
    if (field.type === "boolean") {
      example[field.name] = false;
      continue;
    }
    if (field.type === "number") {
      example[field.name] = 0;
      continue;
    }
    if (field.type === "string") {
      example[field.name] = "";
      continue;
    }
    if (field.type === "string[]") {
      example[field.name] = ["a", "b", "c"];
      continue;
    }
    if (field.type === "number[]") {
      example[field.name] = [1, 2, 3];
      continue;
    }
    if (field.type === "boolean[]") {
      example[field.name] = [false, true];
      continue;
    }
  }
  return example;
}

export function capitalizeFirst(s: string) {
  return s.charAt(0).toUpperCase() + s.slice(1);
}
