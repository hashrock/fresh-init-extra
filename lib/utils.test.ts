// example_test.ts
import { assertSnapshot } from "https://deno.land/std@0.188.0/testing/snapshot.ts";
import { generateExampleFromFields } from "./utils.ts";
import * as kv from "../templates/utils/kv.ts";

Deno.test("generate example json", async function (t): Promise<void> {
  const fields: kv.Field[] = [
    { name: "title", type: "string", nullable: false },
    { name: "body", type: "string", nullable: false },
    { name: "tags", type: "string[]", nullable: false },
    { name: "score", type: "number", nullable: false },
  ];
  const generated = generateExampleFromFields(fields);
  await assertSnapshot(t, generated);
});
