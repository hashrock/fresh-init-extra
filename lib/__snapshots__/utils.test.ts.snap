export const snapshot = {};

snapshot[`generate example json 1`] = `
{
  body: "",
  score: 0,
  tags: [
    "a",
    "b",
    "c",
  ],
  title: "",
}
`;
